Introduce known issues and point people to it (#407)
bencheeorg/benchee: README.md

# Benchee [![Hex Version](https://img.shields.io/hexpm/v/benchee.svg)](https://hex.pm/packages/benchee) [![Hex Docs](https://img.shields.io/badge/docs-hexpm-blue.svg)](https://hexdocs.pm/benchee/) [![CI](https://github.com/bencheeorg/benchee/workflows/CI/badge.svg)](https://github.com/bencheeorg/benchee/actions?query=branch%3Amain) [![Coverage Status](https://coveralls.io/repos/github/bencheeorg/benchee/badge.svg?branch=main)](https://coveralls.io/github/bencheeorg/benchee?branch=main) [![Total Download](https://img.shields.io/hexpm/dt/benchee.svg)](https://hex.pm/packages/benchee) [![License](https://img.shields.io/hexpm/l/benchee.svg)](https://github.com/bencheeorg/benchee/blob/main/LICENSE)

> **If you have benchmarked using elixir 1.14.0-1.16.0-rc.0, please check out [known issues](#known-issues)**

Reviewer: [nit] Maybe would be easier to understand? The double dash made it hard to parse.
```suggestion
**If you have benchmarked using elixir 1.14.0 - 1.16.0-rc.0, please check out [known issues](#known-issues)**
```

This doesn't seem to be too reliable right now, so suggestions and input are very welcome :)

## Known Issues

> There is a known issue affecting elixir version from 1.14.0 to 1.16.0-rc.0: An optimization had been disabled affecting the performance of top level functions. The issue was fixed by re-enabling the optimization in [1.16.0-rc.1](https://github.com/elixir-lang/elixir/blob/v1.16/CHANGELOG.md#v1160-rc1-2023-12-12).

Reviewer: (I'm not sure if this is actually clearer/useful, feel free to ignore 🙂)

```suggestion
There is a known issue affecting elixir versions from 1.14.0 to 1.16.0-rc.0: An optimization had been disabled affecting the performance of functions defined directly in the top level (i.e. outside of any module). The issue was fixed by re-enabling the optimization in [1.16.0-rc.1](https://github.com/elixir-lang/elixir/blob/v1.16/CHANGELOG.md#v1160-rc1-2023-12-12).
```

The issue is best show-cased by the following benchmark where we'd expect ~equal results:

```elixir
list = Enum.to_list(1..10_000)

defmodule Compiled do
  def comprehension(list) do
    for x <- list, rem(x, 2) == 1, do: x + 1
  end
end

Benchee.run(%{
  "module (optimized)" => fn -> Compiled.comprehension(list) end,
  "top_level (non-optimized)" => fn -> for x <- list, rem(x, 2) == 1, do: x + 1 end
})
```

> Which yields ~these results on a non fixed elixir version:

Reviewer: ```suggestion
Which yields ~these results on an affected elixir version:
```

```
Comparison:
module (optimized)              18.24 K
top_level (non-optimized)       11.91 K - 1.53x slower +29.14 μs
```

So, how do you fix it/make sure a benchmark you ran is not affected? All of these work:

> * benchmark on an unaffected/fixed versions of elixir (<= 1.13.4 or >= 1.16.0-rc.1)

Reviewer: ```suggestion
* benchmark on an unaffected/fixed version of elixir (<= 1.13.4 or >= 1.16.0-rc.1)
```
